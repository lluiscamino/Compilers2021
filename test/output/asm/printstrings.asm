.section	__TEXT, __text
	.globl	_main
	.globl	print_uint64
	.globl	print_boolean
	.globl	print_string
	.globl	read_string
	.globl	string_length
	.globl	compare_strings
print_uint64:
	lea 	-1(%rsp), %rsi
	movb	$'\n', (%rsi)
	mov 	$10, %ecx
	movl	%edi, %ebx
	testl	%edi, %edi
	jns 	.print_uint64_positive
	neg 	%edi

.print_uint64_positive:
	mov 	%rdi, %rax

.Ltoascii_digit:
	xor 	%edx, %edx
	div 	%rcx
	add 	$'0', %edx
	dec 	%rsi
	mov 	%dl, (%rsi)
	test	%rax, %rax
	jnz 	.Ltoascii_digit
	testl	%ebx, %ebx
	jns 	.print_uint64_end
	xor 	%edx, %edx
	div 	%rcx
	add 	$'-', %edx
	dec 	%rsi
	mov 	%dl, (%rsi)
	test 	%rax, %rax

.print_uint64_end:
	mov 	$0x02000004, %eax
	mov 	$1, %edi
	mov 	%rsp, %rdx
	sub 	%rsi, %rdx
	syscall
	ret

/**
 * Prints a boolean to stdout
 * Params:
 * - %rbx: Boolean value
 */
print_boolean:
	testl	%ebx, %ebx
	jnz 	.print_boolean_true
	movq	decl_1@GOTPCREL(%rip), %rsi
	jmp 	.print_boolean_end
	.print_boolean_true:
		movq	decl_0@GOTPCREL(%rip), %rsi
	.print_boolean_end:
		call 	print_string
		ret

/**
 * Prints a string to stdout
 * Params:
 * - %rsi: String address
 */
print_string:
	call	string_length
	mov 	$0x02000004, %rax
	mov 	$1, %rdi
	syscall
	ret

/**
 * Reads a string from stdin
 * Params:
 * - %rsi: Address where the string will be saved
 */
read_string:
	mov 	$0x02000003, %rax
	mov 	$0, %rdi
	movq	$140, %rdx
	syscall
	ret

/**
 * Returns a string's length (saves to %rdx)
 * Params:
 * - %rsi: String address
 * https://stackoverflow.com/questions/60482733/how-to-traverse-a-string-in-assembly-until-i-reach-null-strlen-loop
 */
string_length:
	lea 	-1(%rsi), %rdx
	.Lloop:
		inc 	%rdx
		cmpb	$0, (%rdx)
		jne 	.Lloop
	sub 	%rsi, %rdx
	ret

/**
 * Compares two strings (saves result to %rdx)
 * Params:
 * - %rsi: First string address
 * - %rdi: Second string address
 */
compare_strings:
	lea 	-1(%rsi), %rcx
	lea 	-1(%rdi), %rdx
	.Cloop:
		inc 	%rcx
		inc 	%rdx
		cmpb	$0, (%rcx)
		jne 	.compare_strings_continue
		cmpb	$0, (%rdx)
		je  	.compare_strings_true
		.compare_strings_continue:
			movb	(%rcx), %al
			cmpb	%al, (%rdx)
			je  	.Cloop
	.compare_strings_false:
		movq	$0, %rdx
		jmp 	.compare_strings_end
	.compare_strings_true:
		movq	$-1, %rdx
	.compare_strings_end:
		ret

_main:
	push	%rbp
	mov 	%rsp, %rbp
	subq	$0, %rsp
	movq	decl_2@GOTPCREL(%rip), %rsi
	movq	%rbp, (%rsi)
/*t_printString: skip*/
	call	t_main
	mov 	$0x02000001, %rax
	xor 	$0, %rdi
	syscall
t_printString:
/*pmb s0*/
	push	%rbp
	mov 	%rsp, %rbp
	subq	$0, %rsp
/*printString(str)*/
	movq	16(%rbp), %rsi
	call	print_string
/*rtn s0*/
	movq	%rbp, %rsp
	pop 	%rbp
	ret

/*t_main: skip*/
t_main:
/*pmb s1*/
	push	%rbp
	mov 	%rsp, %rbp
	subq	$16, %rsp
/*constString = "Hi! ðŸ˜€\n"*/
	movq	str_3@GOTPCREL(%rip), %rax
	movq	%rax, -8(%rbp)
/*varString = "Hola\n"*/
	movq	str_4@GOTPCREL(%rip), %rax
	movq	%rax, -16(%rbp)
/*printString("Hello, world\n")*/
	movq	str_5@GOTPCREL(%rip), %rsi
	call	print_string
/*printString(constString)*/
	movq	-8(%rbp), %rsi
	call	print_string
/*printString(varString)*/
	movq	-16(%rbp), %rsi
	call	print_string
/*varString = "Adios\n"*/
	movq	str_6@GOTPCREL(%rip), %rax
	movq	%rax, -16(%rbp)
/*printString(varString)*/
	movq	-16(%rbp), %rsi
	call	print_string
/*param_s "749\n"*/
	movq	str_7@GOTPCREL(%rip), %rax
	push	%rax
/*call s0*/
	call	t_printString
/*param_s constString*/
	movq	-8(%rbp), %rax
	push	%rax
/*call s0*/
	call	t_printString
/*rtn s1*/
	movq	%rbp, %rsp
	pop 	%rbp
	ret

.section __DATA, __data
	decl_0: .asciz "true\n"
	decl_1: .asciz "false\n"
	decl_2: .quad 0
	str_3: .asciz "Hi! ðŸ˜€\n"
	str_4: .asciz "Hola\n"
	str_5: .asciz "Hello, world\n"
	str_6: .asciz "Adios\n"
	str_7: .asciz "749\n"
