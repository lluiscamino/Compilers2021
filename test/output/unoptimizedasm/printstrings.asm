.section	__TEXT, __text
	.globl	_main
	.globl	read_string
	.globl	compare_strings
	.globl	print_boolean
	.globl	string_length
	.globl	print_string
	.globl	print_uint64
_main:
	push	%rbp
	mov 	%rsp, %rbp
	subq	$0, %rsp
	movq	decl_2@GOTPCREL(%rip), %rsi
	movq	%rbp, (%rsi)
/*init*/
	call	t_main
	mov 	$0x02000001, %rax
	xor 	$0, %rdi
	syscall
/*t_printString: skip*/
t_printString:
/*pmb s0*/
	push	%rbp
	mov 	%rsp, %rbp
	subq	$0, %rsp
/*printString(str)*/
	movq	16(%rbp), %rsi
	call	print_string
/*rtn s0*/
	movq	%rbp, %rsp
	pop 	%rbp
	ret

/*t_main: skip*/
t_main:
/*pmb s1*/
	push	%rbp
	mov 	%rsp, %rbp
	subq	$56, %rsp
/*t0 = "Hi! ðŸ˜€\n"*/
	movq	str_3@GOTPCREL(%rip), %rax
	movq	%rax, -16(%rbp)
/*constString = t0*/
	movq	-16(%rbp), %rax
	movq	%rax, -8(%rbp)
/*t1 = "Hola\n"*/
	movq	str_4@GOTPCREL(%rip), %rax
	movq	%rax, -32(%rbp)
/*varString = t1*/
	movq	-32(%rbp), %rax
	movq	%rax, -24(%rbp)
/*t2 = "Hello, world\n"*/
	movq	str_5@GOTPCREL(%rip), %rax
	movq	%rax, -40(%rbp)
/*printString(t2)*/
	movq	-40(%rbp), %rsi
	call	print_string
/*printString(constString)*/
	movq	-8(%rbp), %rsi
	call	print_string
/*printString(varString)*/
	movq	-24(%rbp), %rsi
	call	print_string
/*t3 = "Adios\n"*/
	movq	str_6@GOTPCREL(%rip), %rax
	movq	%rax, -48(%rbp)
/*varString = t3*/
	movq	-48(%rbp), %rax
	movq	%rax, -24(%rbp)
/*printString(varString)*/
	movq	-24(%rbp), %rsi
	call	print_string
/*t4 = "749\n"*/
	movq	str_7@GOTPCREL(%rip), %rax
	movq	%rax, -56(%rbp)
/*param t4*/
	movq	-56(%rbp), %rax
	pushq	%rax
/*call s0*/
	call	t_printString
/*param constString*/
	movq	-8(%rbp), %rax
	pushq	%rax
/*call s0*/
	call	t_printString
/*rtn s1*/
	movq	%rbp, %rsp
	pop 	%rbp
	ret

/**
 * Returns a string's length (saves to %rdx)
 * Params:
 * - %rsi: String address
 * https://stackoverflow.com/questions/60482733/how-to-traverse-a-string-in-assembly-until-i-reach-null-strlen-loop
 */
string_length:
	lea 	-1(%rsi), %rdx
	.Lloop:
		inc 	%rdx
		cmpb	$0, (%rdx)
		jne 	.Lloop
	sub 	%rsi, %rdx
	ret

/**
 * Prints a string to stdout
 * Params:
 * - %rsi: String address
 */
print_string:
	call	string_length
	mov 	$0x02000004, %rax
	mov 	$1, %rdi
	syscall
	ret

.section __DATA, __data
	decl_0: .asciz "true\n"
	decl_1: .asciz "false\n"
	decl_2: .quad 0
	str_3: .asciz "Hi! ðŸ˜€\n"
	str_4: .asciz "Hola\n"
	str_5: .asciz "Hello, world\n"
	str_6: .asciz "Adios\n"
	str_7: .asciz "749\n"
