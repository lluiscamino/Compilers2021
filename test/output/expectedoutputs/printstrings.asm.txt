Hello, world
Hi! ğŸ˜€
Hola
Adios
749
Hi! ğŸ˜€